picard/acousticbrainz/__init__.py

```diff
@@ -48,63 +51,97 @@
 from picard.webservice import ratecontrol
 
 
-_acousticbrainz_extractor = None
-_acousticbrainz_extractor_sha = None
 ratecontrol.set_minimum_delay((ACOUSTICBRAINZ_HOST, ACOUSTICBRAINZ_PORT), 1000)
 
+ABExtractorProperties = namedtuple('ABExtractorProperties', ('path', 'version', 'sha', 'mtime_ns'))
 
-def get_extractor(config=None):
-    if not config:
-        config = get_config()
-    extractor_path = config.setting["acousticbrainz_extractor"]
-    if not extractor_path:
-        extractor_path = find_extractor()
-    else:
-        extractor_path = find_executable(extractor_path)
-    return extractor_path
 
+class ABExtractor:
 
-def find_extractor():
-    return find_executable(*EXTRACTOR_NAMES)
+    def __init__(self):
+        self.cache = defaultdict(lambda: None)
+
+    def get(self, config=None):
+        if not config:
+            config = get_config()
+        if not config.setting["use_acousticbrainz"]:
+            log.debug('ABExtractor: AcousticBrainz disabled')
+            return None
+
+        extractor_path = config.setting["acousticbrainz_extractor"]
+        if not extractor_path:
+            extractor_path = find_extractor()
+        else:
+            extractor_path = find_executable(extractor_path)
+        if not extractor_path:
+            log.debug('ABExtractor: cannot find a path to extractor binary')
+            return None
+
+        try:
+            statinfo = os.stat(extractor_path)
+            mtime_ns = statinfo.st_mtime_ns
+        except OSError as exc:
+            log.warning('ABExtractor: cannot stat extractor: %s', exc)
+            return None
+
+        # check if we have this in cache already
+        cached = self.cache[(extractor_path, mtime_ns)]
```

Comment:
Do we really need a list of cache entries? I'd actually go with a single cache item. If path changes it can be re-evaluated. That avoids the question what to do with stale items. Not really an issue, because this usually wouldn't change that often during one run of Picard, but still it seems odd to keep the old entries around.

Reply:
That's not a list but a defaultdict.
Actually it was more because that's simpler to test if there's something in the cache (key is a (path, time) tuple / O(1) check). I thought about potential old entries, but it will happen only in case of changes, so at worse we'll have few entries in the dictionary.
I thought about that of course, but I concluded it wasn't worth bothering.
Without any change to extractor path or binary during runtime, we'll have just one entry (and none at startup).

Reply:
Done in e72c552df0